### ISSUE TITLE
Support for CMS and Frameworks?

### ISSUE THREAD

**Issue by anasik** [TweakStyle/BetaFeedback#67]
This is actually an idea I had back when i first saw this project and pinegrow.com appears to be doing it too.. not sure if they are a competitor for you, but the idea is still implementable.

**Comment by felicienfrancois** [comment]
@anasik thank you for this idea

That's something we are already thinking about but it would be integrated in a different way it think.
Pinegrow is a graphical editor while TweakStyle is a code editor with preview and graphical assist tools.

It would probably be integrated with autocompletion. We're still thinking of it but we have a lot of thing to do before that, so it is not planned yet.

**Comment by anasik** [comment]
I thought the same... TweakStyle cant compete with PineGrow cause they arent even like the same thing.. however STILL there could be some sort of integration for CMS i guess. for example if it is about design.. the user could edit and design it like just another static page and then the tool could automatically generate new files .. 
To be really honest, i dont even know how the CMS integration works in PineGrow. Also, even when considering it'd be wise to keep it limited to CMS as frameworks and stuff dont come in design stuff.. but then theres BootStrap and JQM and stuff.. THAT's what there'll always be room for.

Off Topic: PineGrow seems cheaper.. which might be an attractive factor for users.

**Comment by mikebronner** [comment]
I thought I had brought this up as well back when I first tried TweakStyle out, but I can't find the issues I posted anymore (it very well may have been in emails).

Anyway, might thoughts were along the lines of integrating with backend frameworks, like Laravel Blade, or PHP in general, that would render views by analyzing the pages and assembling all the included files for previewing. This would allow editing and previewing of blade partials without having to manually preview things (otherwise I could still use my IDE, right, there would be no improvement?).

Good discussion, eager to see what comes out of this :) Thanks @felicienfrancois!

**Comment by anasik** [comment]
I've no idea what Blade is.. but as for PHP.. exactly what in PHP can one expect from a tool that focuses on designing? It's kind of hard for me to even picture it.. Which is why I focused more on CMS, and of course all other UI stuff like BootStrap e.t.c.

**Comment by felicienfrancois** [comment]
as far as I understood there are 2 different points in this discussion:
- Framework(s) assist (backend, frontend and UI)
- UI Builder (static and with CMS & UI frameworks support)

TweakStyle is a tool that focus on designing in the sense of styling.
I think none of coding backend or composing frontend is styling BUT both are required to get a frontend to style. So both are included in our plans to make TweakStyle become a complete IDE for web designing/development.

Regarding Framework(s) assist, the plan is to provide this though autompletion and a plugins system. But the way to this is quite long:
- First we'll create an advanced completion framework. By advanced I mean smart, contextual and user friendly. It could be a filterable list (classic one), graphical editor (like color picker), or a filterable completion pane with categories, icons, embedded documentation, ...
- We'll firstly implement this on CSS / SASS / LESS. We could imagine a completion pane for css properties key with categories (text, background, animation, effects, ...) and embedded documentation, then a different completor for each kind of values (colors, dimensions, ...)
- Then we'll make it generic (could be applied to any language, backend or frontend), creates a plugin API, and creates a few example plugins (ex: wordpress, bootstrap).
You could setup your project selecting the completion plugins you need. For example, with a bootstrap plugin, you would get bootstrap classes proposal in HTML, bootstrap functions proposal in javascript. The proposal could be illustrated with doc and preview. With wordpress one you would also get PHP variables and functions proposal.
- We also need a full plugin system and a plugin repository.

Regarding UI Builder, I'm not sure how to integrate this with TweakStyle. Because this is not a code assist. It is more a page creation assist. Maybe the way is:
- First create a page creation assistant, with template selection and framework inclusion (you would be able to select a template and a list of frameworks and it would generate the files.
- Then create a plugin API to able advanced UI to generate pages and creates a first plugin example.
- We also need a full plugin system, a plugin repository and a template repository

There is another feature that is related to this: Edit on preview. I mean the possibility to drag and resize elements in the preview that would updated automatically the code. I'm not sure this is totally feasible in TweakStyle while keeping full control on the code:
- where to write the updated CSS?
- in which language (CSS/LESS/SASS/Stylus) ?
- What about properties shared by several elements ?
- Need to handle all positioning cases ?
- What about all browser polyfill ?
Most graphical editors solve this by removing control over code or restrict possibilities. On top of that most graphical editors don't work (or don't work like expected). That's one of the main reason why I decided to create TweakStyle. To get control over code and not restrict possibilities.

To sum up:
- Framework(s) assist => planned but the way is long
- UI Builder => That's something we could do but the way is long. We'll probably do the first steps to make it possible (page creation assistant and plugin API)
- Edit on preview => We'll probably not do this

**Comment by anasik** [comment]
Yeah FrameWork assist is what i meant actually.. The UI builder thing is like drifting away from what the whole point of this project is. Also, let's say Tweakstyle does get it.. it wont really be able to compete with the big players in the market. No.. TweakStyle is centered on design and designing through code.. 
As for the Edit on Preview thing.. I think it is something TweakStyle could offer.. but then the issues you listed as questions.. perhaps they could be an issue .. so in that case one could make up for this by offering something like the inspect element thing (I opened an issue regarding this a while ago.)
